src/backend/optimizer/plan/createplan.c

```diff
@@ -784,10 +784,14 @@ create_join_plan(PlannerInfo *root, JoinPath *best_path)
 
 	/* CDB: if the join's locus is bottleneck which means the
 	 * join gang only contains one process, so there is no
-	 * risk for motion deadlock.
+	 * risk for motion deadlock. But prefetch_inner may still be necessary in the
+	 * case of HashJoin with SharedScan, the writing part of which is always in
+	 * the inner. So do not override it in the case of HashJoin.
 	 */
-	if (CdbPathLocus_IsBottleneck(best_path->path.locus))
+	if (CdbPathLocus_IsBottleneck(best_path->path.locus) &&
+		!(root->config->gp_cte_sharing && IsA(plan, HashJoin)))
 	{
+		Assert(!partition_selector_created);
```

Comment:
Is it possible to add `else` to `if (partition_selector_created)` instead of this `Assert`?

Reply:
you mean something like this?
```c
if (partition_selector_created)
    ...
else if (CdbPathLocus_IsBottleneck(best_path->path.locus) && 
	    !(root->config->gp_cte_sharing && IsA(plan,HashJoin))){
    ...
}
```
technically yes, but I don't think it makes sense since they are logically unrelated.

Reply:
if we want to be sure that we don't override prefetch_inner in the case of partition selector, then maybe it makes sense to change the order of the conditions?

Reply:
Yes, I do. The logical relation exists, because the assert is added. I suggest to save the order to do not set prefetch_inner two times.